### PR TITLE
[Fix] duplicate code (part 4)

### PR DESCRIPTION
**Pull request by tareknaser360**
Related token: #1264
This PR addresses the issue of duplicate code (part 4). I have added a new function `process_event` that helps avoid code duplication.
If this commit is good, I will continue working on resolving the remaining issues in separate commits.

### CODE REVIEW

File: src/sema/statements.rs [hunk 1]
```diff
@@ -1174,6 +1174,35 @@ fn revert_named_arg(
     }
 }
 
+fn process_event(
```
Comment (xermicus) [inline src/sema/statements.rs:1177]: The name of the function should describe what it does or at least deserves a comment.

File: src/sema/statements.rs [hunk 2]
```diff
@@ -1174,6 +1174,35 @@ fn revert_named_arg(
...
+    ns: &mut Namespace,
+    function_no: &usize,
+    event_no: &usize,
+    event_loc: &pt::Loc,
+    loc: &pt::Loc,
+    cast_args: &[Expression],
+    event_nos: &Vec<usize>,
+    diagnostics: &mut Diagnostics,
+    temp_diagnostics: &Diagnostics,
+) -> Result<Option<Statement>, ()> {
```
Comment (xermicus) [inline src/sema/statements.rs:1188]: I think this function could be a closure.


```
let to_stmt =
        |ns: &mut Namespace, event_no: usize, event_loc: pt::Loc, cast_args: Vec<Expression>| {
            if !ns.functions[function_no].emits_events.contains(&event_no) {
                ns.functions[function_no].emits_events.push(event_no);
            }

            Statement::Emit {
                loc: *loc,
                event_no,
                event_loc,
                args: cast_args,
            }
        };
```

Then you can do:
```
               if matches {
                    return Ok(to_stmt(ns, *event_no, event_loc.clone(), cast_args));
                }
                if event_nos.len() > 1 && diagnostics.extend_non_casting(&errors) {
                    return Err(());
                }
```

**Review by xermicus — CHANGES_REQUESTED** [review]
Just moving some code into a couple helper function should be the last resort. Did you try to refactor the logic of the function `emit_event`? This function has many loops and deep nesting, also it seems to have more duplicate code in it. For example there are two for loops `for event_no in &event_nos {}`

**Comment by seanyoung** [comment]
Your commit also needs a `Signed-off-by: Real Name <email@address>`

File: src/sema/statements.rs (hunk 2 above)
Comment (tareknaser360) [reply src/sema/statements.rs]: Thanks for the review. I called this closure to avoid duplication.

File: src/sema/statements.rs (hunk 1 above)
Comment (tareknaser360) [reply src/sema/statements.rs]: Does the new closure `to_stmt` need a comment?

**Comment by tareknaser360** [comment]
> Your commit also needs a `Signed-off-by: Real Name <email@address>`

Thanks for the feedback.
I signed the new commit.

**Review by seanyoung — APPROVED** [review]

**Review by LucasSte — APPROVED** [review]

**Review by xermicus — APPROVED** [review]

**Pull request merged**